### PR TITLE
benchmark huggingface onnx models

### PR DESCRIPTION
**Pull request by geohotstan**
`LIMIT=50 PYTHONPATH=. python examples/huggingface_onnx.py`
[huggingface_results.json](https://github.com/user-attachments/files/18725239/huggingface_results.json)
didn't do LIMIT=100 since my computer is almost out of storage lol. I'll delete some stuff if we want 100
But current models tested is over 100
```
  "passed": 89,
  "failed": 31,
```

There are still quite a few that fails and I am skipping `SKIPPED_FILES` and `SKIPPED_REPOS` but I think full support 
 might take some more time since there are problems with numerical accuracy for all those quantized models

This can also be ran with `MODELPATH` to  only test a single model.
`MODELPATH=google-bert/bert-base-uncased/model.onnx PYTHONPATH=. python examples/huggingface_onnx.py` 

Any requirements on the output format? And any ideas on how this might be tested in CI? Will it be similar to external_model_benchmark?

### CODE REVIEW

File: examples/benchmark_onnx.py [hunk 1]
```diff
@@ -1,39 +1,87 @@
...
+  # determine value
+  match inp.name:
+    case "input_ids":
+      vocab_size = config.get("text_config", {}).get("vocab_size") or config.get("vocab_size", 50265)
+      val = Tensor.randint(*shape, low=0, high=vocab_size, dtype=dtype)
+    case "attention_mask": val = Tensor.randint(*shape, low=0, high=2, dtype=dtype)
+    case "token_type_ids": val = Tensor.randint(*shape, low=0, high=config.get("type_vocab_size", 2), dtype=dtype)
+    case "image_tensor": val = Tensor.randint(*shape, low=0, high=256, dtype=dtype)
+    case _: val = Tensor.randn(*shape, dtype=dtype).mul(8) if shape else Tensor(None, dtype=dtype)
+  return val.realize()
```
Comment (geohotstan) [inline examples/benchmark_onnx.py:38]: something weird I noticed.
Right now enqueue and total run speeds are similar to master with creation ran in Tensor.

but before [e539692](https://github.com/tinygrad/tinygrad/pull/8493/commits/e539692d9fd87901d2f8daa2ae2053fdd5c988dd)
The values were init-ed with Tensor(numpy.array).realize().
The enqueue and total run speeds was faster (and more stable) than it is currently even though the `input={'0': <Tensor <UOp METAL (1, 3, 224, 224) float (<Ops.VIEW: 20>, <buf real:True device:METAL size:150528 dtype:dtypes.float offset:0>)> on METAL with grad None>}`
was the exact same.

Comment (geohotstan) [reply examples/benchmark_onnx.py:37]: Is this because the extra time spent in local import for .numpy() is attributed towards the total run time? Lemme test it :D

**Comment by geohotstan** [comment]
I'll also add the additional Attention functionality in this PR too.

**Comment by chenyuxyz** [comment]
i would prefer one pr to add onnx ops (can have many), and one pr for the huggingface script

it's generally a good idea for one pr to do one thing (can have multiple sub things), easier to review and easier to revert if needed without reverting too much

**Comment by geohotstan** [comment]
image
[huggingface_results.json](https://github.com/user-attachments/files/18908068/huggingface_results.json)

**Comment by geohotstan** [comment]
@chenyuxyz I think this is ready for you to take a look again.

quick overview:
To start
`LIMIT=100 PYTHONPATH=. python examples/huggingface_onnx_update.py`
this should download the necessary onnx models for the top 100 huggingface repos with "onnx" as a label ranked by 30 days downloads, and produce a `huggingface_repos.yaml` that looks like this
```yaml
repositories:
  FacebookAI/xlm-roberta-large:
    30 day downloads: 122065549
    architecture: XLMRobertaForMaskedLM
    url: https://huggingface.co/FacebookAI/xlm-roberta-large
    path: /Users/zibo/Library/Caches/tinygrad/downloads/huggingface_onnx/models--FacebookAI--xlm-roberta-large/snapshots/c23d21b0620b635a76227c604d44e43a9f0ee389
    models:
    - model: onnx/model.onnx
      size: 0.55MB
  sentence-transformers/all-MiniLM-L6-v2:
    30 day downloads: 98937075
    architecture: BertModel
    url: https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2
    path: /Users/zibo/Library/Caches/tinygrad/downloads/huggingface_onnx/models--sentence-transformers--all-MiniLM-L6-v2/snapshots/fa97f6e7cb1a59073dff9e6b13e2715cf7475ac9
    models:
    - model: onnx/model_O2.onnx
      size: 90.33MB
    - model: onnx/model_O3.onnx
      size: 90.33MB
    - model: onnx/model.onnx
      size: 90.41MB
    - model: onnx/model_O1.onnx
      size: 90.36MB
...
last_updated: '2025-03-06T12:08:57Z'
```
The yaml is just a description of all the models and where they're stored.
the `path` combined with the `model` is the full path where the model is located.

the yaml file is then used to run validation (CI) with
`VALIDATE=huggingface_repos.yaml PYTHONPATH=. python examples/huggingface_onnx_run.py`
If the yaml is created with `examples/huggingface_onnx_update.py` this script, then all the onnx models should be installed and we don't have to worry about if the repos get deleted or updated.

We can refresh the `huggingface_repos.yaml` every... month? and see if all the validation goes through? Delete all the existing folders and download a new set of repos.
We won't get random failures in CI doing it this way.

Let me know if this is an ok idea 🫠 
I'm down to change the "produce yaml -> test using yaml" structure if there's something better.


some notes:
currently with this setup 
`LIMIT=100 PYTHONPATH=. python examples/huggingface_onnx_update.py` all passes.
`LIMIT=100 FILTER_ARCHITECTURE=1 PYTHONPATH=. python examples/huggingface_onnx_update.py` has failures
with `FILTER_ARCHITECTURE=1` we're taken to some uncommon (sometimes hardware specific..) models only downloaded 10,000~ times per 30 days. 
Testing 100 top repos with unique architectures still needs work as does the skipped models and repos.

the get shape and get values in `get_example_inputs` also not ideal.

**Comment by chenyuxyz** [comment]
can you add a dry run flag that still produces yaml but instead of downloading all models just show some metadata about the model / file size that would download

**Comment by geohotstan** [comment]
Added a `--dry_run` flag (decided to use argparser instead of envvar)

Download:
`python examples/huggingface_onnx_download.py --limit 100 --dry_run` to produce a YAML that has the sizes of all the files but doesn't actually downloads the file.
`python examples/huggingface_onnx_download.py --limit 100` Downloads the file with the `path` filled in for where the file is saved in the system.

Run:
`PYTHONPATH=. python examples/huggingface_onnx_run.py --check_ops` to print out a report of ONNX ops used and if we have anything unimplemented, doesn't actually run nor validate.
`PYTHONPATH=. python examples/huggingface_onnx_run.py --validate` runs validation for all the models specified in the YAML
`PYTHONPATH=. python examples/huggingface_onnx_run.py --debug Marqo/marqo-fashionSigLIP/onnx/text_model.onnx` to run single model (can add `--truncate 60`) to validate intermediate results in the model (this truncate mostly works except for very few models)

currently `python examples/huggingface_onnx_download.py --limit 100` then `PYTHONPATH=. python examples/huggingface_onnx_run.py --validate` passes with 192 models validated.

Do you want me to add this to CI this PR?

**Comment by chenyuxyz** [comment]
tested with top10 and looks fine to me. i think we can merge this as is and you can add a separate pr to add something to ci

**Comment by geohotstan** [comment]
> tested with top10 and looks fine to me. i think we can merge this as is and you can add a separate pr to add something to ci

ooooo okok

I made some edits and I think this current one is better!!!! I like this one better :D
Forget the tutorial above. Here's a new one.

STEP 1: MAKE YAML
- `PYTHONPATH=. python extra/huggingface_onnx/collect_metadata.py --limit 100` creates `extra/huggingface_onnx/huggingface_repos.yaml`. 
```
sentence-transformers/all-MiniLM-L6-v2:
    url: https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2
    download_path: null
    files:
    - file: onnx/model.onnx
      size: 90.41MB
...
```
This yaml file has download_path as `null`

STEP 2: DOWNLOAD ONNX MODELS
- `python extra/huggingface_onnx/download_models.py extra/huggingface_onnx/huggingface_repos.yaml` downloads the files into `extra/huggingface_onnx/models/`
```
  sentence-transformers/all-MiniLM-L6-v2:
    url: https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2
    download_path: /Users/zibo/fun/tiny/tinygrad/extra/huggingface_onnx/models/models--sentence-transformers--all-MiniLM-L6-v2/snapshots/c9745ed1d9f207416be6d2e6f8de32d1f16199bf
    files:
    - file: onnx/model.onnx
      size: 90.41MB
```
also the YAML is updated with `download_path`

STEP 3: VALIDATE YAML
- `PYTHONPATH=. python extra/huggingface_onnx/run_models.py extra/huggingface_onnx/huggingface_repos.yaml --validate`


I think doing it like this is much more manageable. Can remove models by commenting out in YAML. Can add models in by copy pasting metadata of other repos into `huggingface_repos.yaml` and running `download_models.py`

**Comment by chenyuxyz** [comment]
cool i ran 3 steps and it all passed (i have 128GB of RAM though)

for the pr that adds this into CI, can you also add a mini README for the steps?

**Comment by chenyuxyz** [comment]
also gitignore these
image
